source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRViewController.mm

```diff
@@ -149,4 +175,212 @@ - (void)render
        [newView.topAnchor constraintEqualToAnchor:view.topAnchor]]];
 }
 
+// Walk through adaptive cards elements recursively and if images/images set/TextBlocks are found process them concurrently
+- (void) addTasksToConcurrentQueue:(std::vector<std::shared_ptr<BaseCardElement>> const &) body
+{
+    for(auto &elem : body)
+    {
+        switch (elem->GetElementType())
+        {
+            case CardElementType::TextBlock:
+            {
+                /// tag a base card element with unique key
+                [self tagBaseCardElement:elem];
+                /// dispatch to concurrent queue
+                std::shared_ptr<TextBlock> txtElem = std::dynamic_pointer_cast<TextBlock>(elem);
+                dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0),
+                    ^{
+                        // MarkDownParser transforms text with MarkDown to a html string
+                        std::shared_ptr<MarkDownParser> markDownParser = std::make_shared<MarkDownParser>(txtElem->GetText().c_str());
+                        NSString *parsedString = [NSString stringWithCString:markDownParser->TransformToHtml().c_str() encoding:NSUTF8StringEncoding];
+
+                        // Font and text size are applied as CSS style by appending it to the html string -- font is hard coded for now
+                        parsedString = [parsedString stringByAppendingString:[NSString stringWithFormat:@"<style>body{font-family: '%@'; font-size:%dpx;}</style>",
+                                                                              @"verdana", [ACRTextBlockRenderer getTextBlockTextSize:txtElem->GetTextSize() withHostConfig:_hostConfig]]];
+                        // Convert html string to NSMutableAttributedString, NSAttributedString knows how to apply html tags
+                        NSData *htmlData = [parsedString dataUsingEncoding:NSUTF16StringEncoding];
+                        NSDictionary *options = @{NSDocumentTypeDocumentAttribute: NSHTMLTextDocumentType};
+                        // Initializing NSMutableAttributedString for HTML rendering is very slow
+                        NSMutableAttributedString *content = [[NSMutableAttributedString alloc] initWithData:htmlData options:options documentAttributes:nil error:nil];
+
+                         dispatch_async(dispatch_get_main_queue(),
+                             ^{ __block UILabel *lab = nil; // generate key for imageMap from image element's id
```

Comment:
>imageMap from image element [](start = 80, length = 27)

Not an image

Reply:
thanks

---
In reply to: [167995671](https://github.com/Microsoft/AdaptiveCards/pull/1124#discussion_r167995671) [](ancestors = 167995671)